### ISSUE TITLE
Allow OME-Zarr for non-microscopy data

### ISSUE THREAD

**Issue by balbasty** [bids-standard/bids-specification#1704]
BIDS accepts ome-zarr data in its microscopy extension, but not in its main MRI-related specification.

In most cases, MRI volumes are small enough that chunked formats do not make sense, but MRIs acquired in ex vivo human brains can become quite large. For example, [these 120 um isotropic MRIs](https://dandiarchive.org/dandiset/000026/draft/files?location=sub-I48%2Fses-MRI%2Fanat&page=1) are about 5GB each. This is particularly problematic for web-based viewing, as most (all?) viewers load the entire volume in GPU memory and have a hard memory limit. The MRI linked above is 1600x1400x640 and I am not sure that niiview would be able to display it. Even if it were, having to download the entire file before showing it makes this impractical.

Could we allow OME-Zarr files in the main spec? 

The main problem I see is that, in its current form, the OME-NGFF specification does not allow storing most of the metadata that live in the nifti header -- most improtantly the qform and sform. As an alternative, we have drafted a very lightweight supplement to OME-Zarr, namely [NIfTI-Zarr](https://github.com/neuroscales/nifti-zarr/tree/draft/0.1), which only requires dumping the nifti header under `.zattrs["nifti"]["base64"]` (using base64 encoding). This makes it very easy to decode the nifti metadata for any library that has access to a base64 decoder, and a nifti parser. We have reference implementations in [python](https://github.com/neuroscales/nifti-zarr/pull/2) and [julia](https://github.com/neuroscales/nifti-zarr/pull/3).

I guess this might be somewhat related to #197

@satra @martindurant @yarikoptic

**Comment by effigies** [comment]
I think this seems like a sensible intermediate between a novel format and the limitations of NIfTI. Is there still a NIfTI working group to get the blessings of?

**Comment by martindurant** [comment]
I was tagged so that I could mention [kerchunk](https://fsspec.github.io/kerchunk/index.html), which could provide a way to read directly from the .nii.gz files in parallel and chunk-wise (chunking limited to the largest dimension) by pretending to be a zarr dataset _without_ rewriting the data at all. This is already possible for uncompressed data, and gzip would take some work ( https://github.com/pauldmccarthy/indexed_gzip/issues/112 ). Having done that, you could create a global virtual zarr dataset over all of the files in the archive. 

If you also want subsampled pyramid (OME) data, you still need to create and store those somewhere, but of course they would be much smaller. The format could be zarr or something else, and kerchunk could present the whole lot as a single zarr dataset.

Note: all this only works in python, but https://observablehq.com/@manzt/ome-tiff-as-filesystemreference (appears to have gone stale) presented a prototype in-browser JS viz of exactly the same thing.

**Comment by martindurant** [comment]
>  dumping the nifti header under .zattrs["nifti"]["base64"]

Mild comment: this seems an odd choice to me (but I don't know the domain). The nice thing about JSON metadata is that you can trivially read it, so why not (also?) include the fields, e.g., as interpreted by nibabel `dict(nibabel.nifti1.Nifti1Header)`?

**Comment by yarikoptic** [comment]
Agree with @martindurant on unclear reason for an attempt to preserve nifti header in its original binary form. I can only guess that rationale was to facilitate 1-to-1 binary roundtrip nifti-zarr-nifti. But I do not think that it is that much needed or desired. JSON choice there would have been much better.  Continuing on

> OME-NGFF specification does not allow storing most of the metadata that live in the nifti header -- most improtantly the qform and sform

I wonder if NIfTI-Zarr considered adopting superset of NIfTI + .json sidecar fields defined by BIDS already to be included within OME-Zarr?
From the other side -- shouldn't we in BIDS converge on harmonization of metadata in sidecar JSON file to cover also metadata people rely on getting directly from NIfTI (ie sform/qfrorm, AcquisitionMatrixExtent, ...)? 

Also attn @matthew-brett as he was into "new imaging format" considerations.  

Overall it sounds like a separate and large issue to discuss so we might want a dedicated another issue to it.  But it also feels like a prerequisite to have a complete answer to this one. As @balbasty noted that OME-Zarr lacks needed metadata, we apparently even lack it in sidecar files, and it is unlikely that we would accept some *ad-hoc* (not "agreed upon or already widely used") solution within OME-Zarr.

**Comment by martindurant** [comment]
> OME-Zarr lacks needed metadata

Do you mean it lacks *required* fields, or that the metadata structure doesn't allow for the kind of information you want to preserve? I find in other fields (particularly earth observation and climatology), a huge amount of complex metadata is stored in zarrs.

**Comment by satra** [comment]
@martindurant and @yarikoptic - the expanded form of the binary header in json form is also included in the zarr metadata in @balbasty example.

**Comment by effigies** [comment]
Another approach to the B64-in-JSON encoding for the header would be to create a `nifti` array that is just the literal NIfTI header:

```python
import nibabel as nb
import numpy as np
import zarr

img = nb.Nifti1Image(np.zeros((256,256,256), dtype='f4'), np.diag((2, 2, 2, 1)))

root = zarr.group(store='/tmp/img.nii.zarr')
root.array(name='0', data=img.dataobj)
root.array(name='nifti', data=img.header.binaryblock)

rt_header = nb.Nifti1Header(binaryblock=np.asanyarray(nifti).tobytes())
round_trip = nb.Nifti1Image(
    root['0'],
    affine=rt_header.get_best_affine(),
    header=rt_header,
)
```

Playing with a real file, I was able to read the header with `mri_info`:

```
❯ mri_info /tmp/nii.zarr/nifti/0
Volume information for /tmp/nii.zarr/nifti/0
          type: nii
    dimensions: 208 x 300 x 320
   voxel sizes: 0.800000, 0.800000, 0.800000
          type: FLOAT (3)
           fov: 166.400
           dof: 1
        xstart: -83.2, xend: 83.2
        ystart: -120.0, yend: 120.0
        zstart: -128.0, zend: 128.0
            TR: 2400.00 msec, TE: 0.00 msec, TI: 0.00 msec, flip angle: 0.00 degrees
       nframes: 1
       PhEncDir: UNKNOWN
       FieldStrength: 0.000000
ras xform present
    xform info: x_r =   0.9981, y_r =   0.0621, z_r =   0.0057, c_r =    -0.9937
              : x_a =  -0.0622, y_a =   0.9980, z_a =   0.0072, c_a =    18.6450
              : x_s =  -0.0052, y_s =  -0.0075, z_s =   1.0000, c_s =     4.2302
Orientation   : RAS
Primary Slice Direction: axial

voxel to ras transform:
                0.7984   0.0497   0.0045   -92.2098
               -0.0497   0.7984   0.0057   -96.8662
               -0.0042  -0.0060   0.8000  -122.4309
                0.0000   0.0000   0.0000     1.0000

voxel-to-ras determinant 0.512

ras to voxel transform:
                1.2476  -0.0777  -0.0065   106.7157
                0.0776   1.2476  -0.0094   126.8562
                0.0071   0.0090   1.2499   154.5524
                0.0000   0.0000   0.0000     1.0000
```

Any tool that can currently parse a NIfTI header should be able to work with this structure.

**Comment by martindurant** [comment]
@effigies , that's quite clever :)

**Comment by yarikoptic** [comment]
indeed quite sneaky! example needs little fixing though

```shell
❯ python tryniftihdr.py
Traceback (most recent call last):
  File "/tmp/tryniftihdr.py", line 11, in <module>
    rt_header = nb.Nifti1Header(binaryblock=np.asanyarray(nifti).tobytes())
                                                          ^^^^^
NameError: name 'nifti' is not defined
```

**Comment by effigies** [comment]
Fixed. Though the goal was to be lazy, not sneaky.

**Comment by yarikoptic** [comment]
Being sneaky while being lazy is a true super-power! ;)

**Comment by yarikoptic** [comment]
@satra I re-read it and neither I found any update to it since I originally read it years back, nor answers to my questions. I feel your pain for your particular use case, but it doesn't make it common (yet at least). Thus I was trying to see if we could identify compromise. But it might be useful to first discover what practical benefits you are aiming for from BIDS as seeing most important: e.g. validation of structure? or metadata? encouraging downstream to BIDS neuroimaging tools to support Nifti-Zarr?  Asking because it might help to pave the pragmatic path forward here.

**Comment by martindurant** [comment]
Thanks for pinging me on here. I am happy to see progress, and anything that makes your data readable in NG without complete download, that's the point. On the conversation about exactly how the metadata should be stored - I don't know. There isn't even that much of it, but you need both sides to agree, so it's a social roblem more than a technical one.

On a side note, zarr has had something of a renaissance in the last two years. I don't know how well that transfers to OME or neuro/microscopy data in general.

**Comment by kabilar** [comment]
> On a side note, zarr has had something of a renaissance in the last two years. I don't know how well that transfers to OME or neuro/microscopy data in general.

The OME and neuro/microscopy worlds are quite engaged with OME-Zarr.

The OME-Zarr team had a challenge in 2024 and contributors generated 422 TB of public biomedical data in Zarr v3.  See https://ome.github.io/ome2024-ngff-challenge/ and https://github.com/ome/ome2024-ngff-challenge.  And the team is working towards OME-Zarr v1.0.  cc @joshmoore

I can't speak too broadly, but several of the NIH BRAIN CONNECTS projects (https://www.brain-connects.org/) are all in with OME-Zarr.

**Comment by kabilar** [comment]
Hi @yarikoptic et al., following up here to see if you would be okay with the change proposed below:

> Perhaps we don't impose OME-Zarr but provide the option to use it. For the [Imaging files](https://bids-specification.readthedocs.io/en/stable/common-principles.html#imaging-files) section, perhaps it could be changed as shown below.
> 
> Original:
> 
> ```
> All imaging data MUST be stored using the NIfTI file format.
> ...
> ```
> 
> Suggestion:
> 
> ```
> Imaging data SHOULD be stored using the NIfTI file format.  Large imaging data MAY be stored using the [OME-Zarr](https://ngff.openmicroscopy.org/) file format.
> ...
> ```
> 
> We could provide further guidance as to why one may want to store imaging data in OME-Zarr. And we could provide specifics about how to store the metadata as [@balbasty](https://github.com/balbasty) described above.

Thanks.

**Comment by yarikoptic** [comment]
overall, I think it is worth initiating a PR or might even be a BEP. FWIW I recall that we did discuss Zarr a few years back at BIDS maintainers meeting as a generic store for multidimensional data. So it might be worth mentioning that.

**Comment by kabilar** [comment]
Thanks Yarik.  I have filed https://github.com/bids-standard/bids-specification/pull/2392.

**Comment by SylvainTakerkart** [comment]
Hi everyone, I'm browsing the BIDS spec repo issues (looking for something else) and arrived here...

By curiosity, I have one (probably naive) question : why are you considering ome.zarr here and not simply (and more generically) zarr?
